### ISSUE TITLE
Array index ranges

### ISSUE THREAD

**Issue by LmpOSX** [kaitai-io/kaitai_struct#680]
Hi,

  I wonder if is there a way to get an array slice using index ranges like a[3:5], similar to python index ranges. I tried but couldn't get it work, perhaps I missed something.

  It is something similar implemented in Kaitai string with substring method,  but it would be good to have similar feature for arrays (either byte o true array).

Regards.

**Comment by generalmimon** [comment]
> I wonder if is there a way to get an array slice using index ranges like a[3:5]

There isn't, because nobody didn't seem to need it. Could you provide some real example format or the situation where you need it?

**Comment by dgelessus** [comment]
For reference, #282 also suggests the same feature.

As for real use cases, there are two that come to mind:

* On byte arrays you'd use slices for the same reasons you'd use `substring` on strings - although according to a GitHub search through kaitai_struct_formats, no one seems to be using `substring` either :) I could've used byte array slices in some places as a crude replacement for a "starts with" operation, but I've [worked around](https://github.com/kaitai-io/kaitai_struct_formats/blob/6bb4a3d31fbaaed26c16a2e3b3f15acf3920cdd7/archive/ar/ar_bsd.ksy#L56-L74) that limitation by re-parsing the byte array using an instance.
* On regular arrays you would use it if some elements at the start/end of the array should be treated differently for whatever reason. This is another thing you can sometimes work around by restructuring your KSY, but it can get annoying if (for example) the first element is only *sometimes* special, and you know this only after you've already parsed it.

**Comment by GreyCat** [comment]
Closing this one as duplicate of #282. Please add to existing discussion there.

**Comment by LmpOSX** [comment]
> > I wonder if is there a way to get an array slice using index ranges like a[3:5]
> 
> There isn't, because it didn't seem like anyone needed it yet. Could you provide some real example format or the situation where you need it?

Well, I need it for example to extract encoded values.
I have some fields, defined as 7 bytes read as type u1 / size 7. Then there I have various values encoded, first is 3 bytes, second and third 2. I need a[:3],a[3:5],a[5:7] on instances/values calculations. Right now I cast to string and use substring method then cast back to the type needed.

**Comment by GreyCat** [comment]
Out of curiosity, why can't / don't you use a substream?

```yaml
seq:
  - id: my_7_bytes
    size: 7
    type: dummy
instances:
  first:
    io: my_7_bytes._io
    pos: 0
    size: 4
    type: ...
  second:
    io: my_7_bytes._io
    pos: 4
    size: 2
    type: ...
  
types:
  dummy: {}

**Comment by LmpOSX** [comment]
I recall trying it , but didn't compile or something, I'think I needed it as a parameter, which doesn't have io,  but it was on my first steps using kaitai. So I will try again because it seems great and will comeback with any progress.
